Fix the Queue to call API endpoints without internal HTTP routing

whitphx opened the pull request:
## Description

Replace the API calls from the queue via HTTP with direct function calls.
Dispatching HTTP requests from a server to the same server violates the concept that the HTTP API endpoints are an interface between inside and outside the server, and it also introduces unnecessary implicit dependency to the underlying HTTP server. In practice, it makes a problem when porting the server into the Wasm env (Gradio-lite).

This PR can be the basis of #5124, replacing its commit emulating the HTTP req/res without actual HTTP communications (67c432d)

At this point, this PR is just a draft showing the concept of this kind of refactoring, and there are rooms for improvements.

Roughly, the code has been changed as follows:
* In `routes.py`, `run_predict()` was inlined into `predict()` (the endpoint handler of `/api/{api_name}`) because `run_predict()` was only used there and looked like this function separation is not so reasonable.
* Copy the content of `predict()` into `queueing.Queue.call_prediction()` replacing the original HTTP call. As a result, `call_prediction` returns the result data as dict directly instead of HTTP response object.
* Extract some code blocks into `route_utils.py`.

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

abidlabs commented on gradio/queueing.py:
```diff
@@ -364,15 +363,73 @@ async def call_prediction(self, events: list[Event], batch: bool):
         )
-        return response
+
+        session_state, iterators = route_utils.restore_session_state(app=app, body=body)
+
+        fn_index = body.fn_index
+
+        event_id = getattr(body, "event_id", None)
+        raw_input = body.data
+
+        dependency = app.get_blocks().dependencies[fn_index_inferred]
+        target = dependency["targets"][0] if len(dependency["targets"]) else None
+        event_data = EventData(
+            app.get_blocks().blocks.get(target) if target else None,
+            body.event_data,
+        )
+        batch = dependency["batch"]
+        if not (body.batched) and batch:
+            raw_input = [raw_input]
+        try:
+            with utils.MatplotlibBackendMananger():
+                output = await app.get_blocks().process_api(
+                    fn_index=fn_index_inferred,
+                    inputs=raw_input,
+                    request=gr_request,
+                    state=session_state,
+                    iterators=iterators,
+                    event_id=event_id,
+                    event_data=event_data,
+                )
+            iterator = output.pop("iterator", None)
+            if hasattr(body, "session_hash"):
+                app.iterators[body.session_hash][fn_index] = iterator
+            if isinstance(output, Error):
+                output = default_response_class(output)  # Added
+                raise output
```
Why are we duplicating all of the logic from inside the `gr.routes.predict()` method instead of just calling the `gr.routes.predict()` method directly?

whitphx replied:
Sorry I just left it undone but it must be extracted to a shared method as you wrote.

whitphx replied:
Thank you!
I gave a try on it playing around the code a bit, but I thought the first approach looked better then. Let me choose the first way for now.
The reasons are:
* Obtaining the endpoint handler from the router object is not documented either in FastAPI or Starlette documents, so it seems not a public API.
* Starlette does it for routing like [this](https://github.com/encode/starlette/blob/a8b8856ce393a82ab4a714131085dbc4d658e34d/starlette/routing.py#L740-L758C19) internally (FastAPI delegates the routing to Starlette). Doing it by ourselves would be reimplementation of the framework internals, and it can be seen the same as [another approach I suggested emulating the ASGI protocol](https://github.com/gradio-app/gradio/pull/5124/commits/67c432dedf240ff07fe5d96a9234daa84d1ec5e0), which we considered not natural, in that they are imitation of framework's roles.

abidlabs commented:
Nice @whitphx! I like this approach, and I agree that its better to pass the FastAPI app into the `Queue` and call the app's predict route directly rather than making an HTTP request

My main suggestion -- can't we literally call the FastAPI app's `predict()` function and pass in these parameters:

```py
        async def predict(
            api_name: str,
            body: PredictBody,
            request: fastapi.Request,
            username: str = Depends(get_current_user),
        ):
```

instead of duplicating the logic that is currently in the predict() method (or even instead of moving that logic elsewhere in a shared function). It's possible to get access FastAPI app's methods using the `router` object and it feels cleaner than this approach. 

Let me also cc @freddyaboulton and @aliabid94 for their thoughts

freddyaboulton commented on gradio/routes.py:
```diff
@@ -400,33 +395,32 @@ async def reset_iterator(body: ResetBody):
+                app=app, api_name=api_name, body=body
+            )
+
+            gr_request = route_utils.compile_gr_request(
```
Very clever @whitphx ! The one thing that jumped out at me is that this breaks `api_open=False`. Some app developers don't want requests to skip the queue so I don't think we should break that functionality.

whitphx replied:
Thank you, and sorry that's right, I will revert it.

whitphx replied:
Do you think this change on the if block makes sense?

Before:
```python
            if (
                not app.get_blocks().api_open
                and app.get_blocks().queue_enabled_for_fn(fn_index_inferred)
                and f"Bearer {app.queue_token}" != request.headers.get("Authorization")
            ):
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Not authorized to skip the queue",
                )
```

After:
```python
            if not app.get_blocks().api_open:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                )
```


https://github.com/gradio-app/gradio/pull/5165/files#diff-324a7165f5d5a8823a28b76f5653fa45f32c8144c82b2e528882c97c7eae534fR409-R412

freddyaboulton replied:
I think we still need to check if the queue is enabled for the function. Otherwise I think this would work. Also, I think 401 code is still appropriate here, no?

whitphx replied:
> I think we still need to check if the queue is enabled for the function.

Ahh, thank you, I misunderstood this if-condition. This API endpoint can be closed only when the queue is enabled as another gateway to access the functionality.

>  Also, I think 401 code is still appropriate here, no?

Actually I don't have a solid opinion on it and 401 would be ok.

Just let me dump my thoughts:
I thought 401 would implies that there is still possibility for the client to make auth successful by providing the correct credentials, and the failure is due to client's responsibility. However, now the availability of this endpoint is controlled completely by the server-side configs (queue and `api_open` flag), so 404 may make more sense.

whitphx commented:
@abidlabs 
> My main suggestion -- can't we literally call the FastAPI app's predict() function and pass in these parameters:

Thanks, I will try it :D

whitphx commented:
@abidlabs 
>  It's possible to get access FastAPI app's methods using the `router` object 

Could you please provide an example?
Would it be something like this? https://github.com/encode/starlette/blob/a8b8856ce393a82ab4a714131085dbc4d658e34d/starlette/routing.py#L740-L758C19

abidlabs commented:
I'm using to seeing something like this:

```py
predict_route = next(route for route in app.router.routes if route.path == "/api")
predict_endpoint = predict_route.endpoint

# call endpoint
await predict_endpoint(*params)
```

but yeah the other approach you mentioned could work too

whitphx commented:
@abidlabs @aliabid94 @freddyaboulton Can you take a glance at this draft PR again as I added some mods after your comments.

The original intention was just to extract the content of the `predict` endpoint handler into a reusable function, to call it from the queue directly, and to remove the HTTP calls from the queue,
but during several tries on it, I first modified `predict()` extracting its code into a few methods so that I can understand the semantics in that function, which was kind of refactoring. As a result, now the queue calls these methods instead of a single handler.
There is still an option to extract the entire contents of `predict()` and `Queue.call_prediction` into a single shared function, but I think it might be too much.

---
TODO:
* Do the same thing for the `/reset` endpoint which is also called from the queue.

abidlabs commented:
Thanks @whitphx, this looks much, much cleaner. I resolved the conflicts, bringing in some of the recent changes around steraming into this branch

Let me know when this PR is out of draft and you'd like a full review!

whitphx commented:
Looks like some unresolved conflicts remain in the last 3 commits.
Let me reset them and merge the main again with conflict resolution. Just in case, I backed up the current branch status to #5358.

whitphx commented on gradio/route_utils.py:
```diff
@@ -0,0 +1,249 @@
+
+    def __init__(
+        self,
+        request: Union[fastapi.Request, None] = None,
```
This PEP604 syntax (`X | Y` for `Union[X, Y]`) is not supported by Python<3.10 and the following error has been raised in the CI on this branch actually,
but looks like the original code (`routes.py`) has passed the tests on the main branch... why?
```
No configuration file found.
pyproject.toml file found at /home/runner/work/gradio/gradio.
Loading pyproject.toml file at /home/runner/work/gradio/gradio/pyproject.toml
Assuming Python version 3.8
Assuming Python platform Linux
Searching for source files
Found 79 source files
pyright 1.1.305
/home/runner/work/gradio/gradio/gradio/route_utils.py
  /home/runner/work/gradio/gradio/gradio/route_utils.py:89:23 - error: Alternative syntax for unions requires Python 3.10 or newer
  /home/runner/work/gradio/gradio/gradio/route_utils.py:88:34 - error: Alternative syntax for unions requires Python 3.10 or newer
  /home/runner/work/gradio/gradio/gradio/route_utils.py:[19](https://github.com/gradio-app/gradio/actions/runs/5996574137/job/16261357539?pr=5165#step:14:20)2:[25](https://github.com/gradio-app/gradio/actions/runs/5996574137/job/16261357539?pr=5165#step:14:26) - error: Alternative syntax for unions requires Python 3.10 or newer
3 errors, 0 warnings, 0 informations
```

abidlabs replied:
You just need to add:

```py
from __future__ import annotations
```

at the top of your file

whitphx replied:
omg! Thanks :D

whitphx commented:
Almost working, but chatinterface_echo example.
Let me take a bit longer for debugging it...

whitphx commented on gradio/routes.py:
```diff
@@ -728,124 +635,14 @@ def get_types(cls_set: List[Type]):
     app: fastapi.FastAPI,
     blocks: gradio.Blocks,
     path: str,
-    gradio_api_url: str | None = None,
```
Is this breaking change OK?

whitphx replied:
I reverted it because this made the CI fail, maybe at building the document.
https://github.com/gradio-app/gradio/actions/runs/6000343454/job/16272211763?pr=5165

whitphx commented on gradio/routes.py:
```diff
@@ -875,8 +673,6 @@ def read_main():
     async def start_queue():
         if gradio_app.get_blocks().enable_queue:
-            if gradio_api_url:
-                gradio_app.get_blocks()._queue.set_url(gradio_api_url)
```
Now the `gradio_api_url` argument of `mount_gradio_app` is no longer used,
but I kept it for now because its deletion makes the CI fail, probably at building the document.
What's the best way?

```python
def mount_gradio_app(
    app: fastapi.FastAPI,
    blocks: gradio.Blocks,
    path: str,
    gradio_api_url: str | None = None,
    app_kwargs: dict[str, Any] | None = None,
)
```

whitphx replied:
Got it! -> https://github.com/gradio-app/gradio/pull/5165/commits/29a74c6634bcf93dc11a1261f06daa17b16afd40

abidlabs commented on gradio/queueing.py:
```diff
@@ -35,7 +37,7 @@ def __init__(
         self.data: PredictBody | None = None
         self.lost_connection_time: float | None = None
-        self.token: str | None = None
+        self.username: Optional[str] = None
```
For consistency with the rest of the codebase:
```suggestion
        self.username: str | None = None
```

abidlabs commented on gradio/utils.py:
```diff
@@ -350,225 +345,6 @@ async def async_iteration(iterator):
     return await iterator.__anext__()
 
 
-class AsyncRequest:
```
Yay! Never liked this class

abidlabs commented:
Thanks again @whitphx! Changes look good. The only issue I noticed is that `gr.Request` isn't working with the queue anymore. Here's a simple repro:

```py
import gradio as gr

def predict(text, request: gr.Request):
    headers = request.headers
    return {
        "headers": headers,
    }

gr.Interface(predict, "text", "json").queue().launch()
```

I think the `Obj` class doesn't seem to be working properly, in that its not correctly convert dictionaries into objects.

whitphx commented:
> The only issue I noticed is that gr.Request isn't working with the queue anymore. 

Thank you, I missed it 😭 
Fixed! 👉 https://github.com/gradio-app/gradio/pull/5165/commits/025d47d00a82125b83c1bd9878a03de51adfe380

whitphx commented:
I created a new PR and confirmed all CI jobs passed:  #5377

whitphx commented:
> The only issue I noticed is that gr.Request isn't working with the queue anymore. 

Thank you, I missed it 😭 
Fixed! 👉 https://github.com/gradio-app/gradio/pull/5165/commits/9f51ac6a7f98b7231a7fa7a550c9a7ebff55ac34

I addressed all your comments. Hope you give a review again, thanks!

abidlabs commented:
Thank you @whitphx will take another look!

abidlabs commented:
LGTM! Awesome PR @whitphx :)

whitphx commented on gradio/route_utils.py:
```diff
@@ -0,0 +1,248 @@
+):
+    # If this fn_index cancels jobs, then the only input we need is the
+    # current session hash
+    if app.get_blocks().dependencies[fn_index_inferred]["cancels"]:
```
Thank you! This is a critical point. I will take a look into it...

whitphx replied:
### Reason
This line is what cancels the task and the cause of that `CancelledError`:
https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/routes.py#L655

The cancelled task here is:
1. At first, set at `Queue.start_processing()`.
   https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/queueing.py#L146
2. Then repeatedly renewed just after cancelling the existing one with the same task name:
   https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/routes.py#L658-L660

and even in the main branch code, the `CancelledError` has been raised when the task is cancelled, but there was no `try-except` block to catch it.
(For example, you can catch the error by adding `except asyncio.CancelledError` to the following try-block:
https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/queueing.py#L391
)

What changed is this `except BaseException` handler:
https://github.com/gradio-app/gradio/pull/5165/files#diff-cab37b7b6ce45220156832fd4ea1597bedcb26f5cfc43ab5c678fc8a586a0232R398
I just copied it [from `routes.py`](https://github.com/gradio-app/gradio/blob/3d66e61d641da8ca2a7d10c545c7dc0139697f00/gradio/routes.py#L538) to do the same thing in `queueing.py`, but it accidentally worked as a handler for the `CancelledError` and prints the stack trace with `traceback.print_exc()`.

### Solution
To ignore the `CancelledError` keeping the same behavior as before for other exceptions, we can just add `except asyncio.CancelledError` to the `try-except` block:
```diff
                 iterators=iterators,
                 fn_index_inferred=fn_index_inferred,
             )
+        except asyncio.CancelledError:
+            raise
         except BaseException as error:
             show_error = app.get_blocks().show_error or isinstance(error, Error)
             traceback.print_exc()
```

However, as a general practice, `except Exception` is preferred to `except BaseException`:
> Exceptions which are not subclasses of Exception are not typically handled, because they are used to indicate that the program should terminate.
> https://docs.python.org/3/tutorial/errors.html#handling-exceptions

So replacing `except BaseException` with `except Exception` looks better if we recall the reason why `BaseException` is used here and can confirm that it's ok to replace it.

whitphx replied:
@freddyaboulton @abidlabs I pushed the fix: 0b2c42c.
However, as commented above, replacing `except BaseException` with `except Exception` might be better instead of this commit, if we can judge it's ok knowing the reason of having been using `BaseException`.
Do you have any idea?

abidlabs replied:
Should be fine to replace with `except Exception` -- any objections @freddyaboulton?

freddyaboulton replied:
Thanks @abidlabs and @whitphx !

whitphx commented on gradio/queueing.py:
```diff
@@ -359,32 +349,69 @@ def get_request_params(self, websocket: fastapi.WebSocket) -> dict[str, Any]:
+            app=app, api_name=api_name, body=body
+        )
+
+        gr_request = route_utils.compile_gr_request(
```
Thank you! You are right :D
* I moved `restore_session_state` into `call_process_api`.
* I'm still wondering if I should move `traceback.print_exc()` into it, because it's more interface-like code printing output, but not a logic, so putting it on the current position also makes sense, for example `predict()` in `routes.py` that is an endpoint handler or the I/O layer.

abidlabs commented:
Tested and everything looks good to me, thank you again @whitphx for the great PR!
